crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -1,83 +1,858 @@
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Read {
+    owner: Option<String>,
+    name: String,
+}
+
+impl State for Read {
+    fn owner(&self) -> Option<&str> {
+        self.owner.as_deref()
     }
 
-    pub async fn uninstall(&self, _packages: &[FloxPackage]) {
-        todo!()
+    fn name(&self) -> &str {
+        &self.name
     }
+}
 
-    pub async fn upgrade(&self, _packages: &[FloxPackage]) {
-        todo!()
+#[derive(Debug)]
+pub struct Modify {
+    /// The original [Read]-Only Environment
+    origin: Environment<Read>,
+}
+
+impl State for Modify {
+    fn owner(&self) -> Option<&str> {
+        self.origin.state.owner()
     }
 
-    pub fn named(self) -> Option<floxmeta::environment::Environment<'flox, Git, ReadOnly<Git>>> {
-        match self {
-            CommonEnvironment::Named(n) => Some(n),
-            CommonEnvironment::Project(_) => None,
+    fn name(&self) -> &str {
+        self.origin.state.name()
+    }
+}
+
+/// Object tracking an environment collection (`.flox`) folder
+pub struct DotFloxDir {
+    path: PathBuf,
+}
+
+impl DotFloxDir {
+    /// Find the closest `.flox` starting with `current_dir`
+    /// and looking up ancestor directories until `/`
+    pub fn discover(current_dir: impl AsRef<Path>) -> Result<Self, EnvironmentError2> {
+        let dot_flox = current_dir
+            .as_ref()
+            .ancestors()
+            .find(|ancestor| ancestor.join(".flox").exists())
+            .ok_or(EnvironmentError2::NoDotFloxFound)?;
+
+        Self::open(dot_flox)
+    }
+
+    /// Open a `.flox` directory in a specific path.
+    ///
+    /// The method expects a path that _contains_ a `.flox` directory!
+    pub fn open(path: impl AsRef<Path>) -> Result<Self, EnvironmentError2> {
+        let dot_flox = path.as_ref().join(".flox");
+        if !dot_flox.exists() {
+            Err(EnvironmentError2::NoDotFloxFound)?
+        }
+        if !dot_flox.is_dir() {
+            Err(EnvironmentError2::DotFloxNotADirectory)?
+        }
+
+        let path = dot_flox
+            .canonicalize()
+            .map_err(EnvironmentError2::DotFloxCanonicalize)?;
+
+        Ok(Self { path })
+    }
+
+    /// Create a new `.flox` directory in a specific path or open it if it exists.
+    ///
+    /// The method creates or opens a `.flox` directory _contained_ within `path`!
+    pub fn new(path: impl AsRef<Path>) -> Result<DotFloxDir, EnvironmentError2> {
+        if !path.as_ref().join(".flox").exists() {
+            std::fs::create_dir(path.as_ref().join(".flox"))
+                .map_err(EnvironmentError2::CreateDotFlox)?;
+        }
+        Self::open(path)
+    }
+
+    /// Get the basepath of the `.flox` dir
+    ///
+    /// This is the path that _contains_ the `.flox` dir!
+    pub fn path(&self) -> &Path {
+        &self.path
+    }
+
+    /// List the environments defined in the `.flox` dir.
+    pub fn environments(&self) -> Result<Vec<Environment<Read>>, EnvironmentError2> {
```

Comment:
I believe we plan to support only one environment per directory. I'm not opposed to this implementation, but I think we should be careful to not waste time future proofing. There's a few things throughout that we should change throughout if we're going to enforce a single environment.

But I don't think a `.flox` directory should have any concept of owner (unless it's a link to FLOX_META, but I think that's something different)

Reply:
This work was roughly based on the project envs we used to have, thus .flox here can hold multiple environments (or links, but i digress).
The spec mentions "needs engineering" and i remember we were discussing that point. however, a single env per directory would make the `DotFloxDir` struct either simpler or close to unnecessary (pending link handling...)